### ISSUE TITLE
[Regression] IDE0370: False positive on unconstrained generic on .NET 11 Preview1

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#82483]
**Version Used**: 
.NET 11 Preview 1 with VS Insiders 2026 [11513.90]

This is a regression. The problem does not occur using stable .NET 10 and stable VS 2026.

**Steps to Reproduce**:
A suggestion is offered to convert:
```c#
var id = (TId)operation.Resource.GetTypedId();
await _service.DeleteAsync(id! /* SUGGESTION HERE */, cancellationToken);
```
to:
```c#
var id = (TId)operation.Resource.GetTypedId();
await _service.DeleteAsync(id, cancellationToken);
```

In the code above, `TId` is an unconstrained generic type (so it can hold a reference or value type).
The called method is defined as:
```c#
Task DeleteAsync([DisallowNull] TId id, CancellationToken cancellationToken);
```

After applying the fix, ReSharper reports the following error:

<img width="741" height="134" alt="Image" src="https://github.com/user-attachments/assets/0613f5c1-82b8-48f8-9cb8-7330f70944f3" />

**Diagnostic Id**:
IDE0370: Suppression is unnecessary.

**Expected Behavior**:
No suggestion to change the existing code.

**Actual Behavior**:
A suggestion is offered. We've increased the severity to Warning in Release builds. Combined with warnaserror, this breaks the build.

**Comment by CyrusNajmabadi** [comment]
> After applying the fix, ReSharper reports the following error:

Resharper issues should be raised with them. 

The analyzer works by removing the `!` and seeing if Roslyn compiler reports a warning. If not. The `!` can be removed.

**Comment by bkoelman** [comment]
Could this be a compiler bug, then? It seems to me that ReSharper is actually correct, unless the compiler rules have changed. Here's a minimal repro:

Project file:
```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net11.0</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
		<AnalysisMode>Recommended</AnalysisMode>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
		<MSBuildTreatWarningsAsErrors>true</MSBuildTreatWarningsAsErrors>
		<SuppressNETCoreSdkPreviewMessage>true</SuppressNETCoreSdkPreviewMessage>
		<NoWarn>$(NoWarn);IDE0005;IDE0008;IDE0290;CA1822</NoWarn>
	</PropertyGroup>
</Project>
```

`.editorconfig`:
```
[*.cs]

# CA1515: Consider making public types internal
dotnet_diagnostic.IDE0370.severity = error
```

`Program.cs`:
```c#
using System.Diagnostics.CodeAnalysis;

var example = new Example<int?>(null);
example.M();

internal sealed class Example<TId>
{
    private readonly TId _id;

    public Example(TId id)
    {
        _id = id;
    }

    public void M()
    {
        var id = (TId)GetId();
        Delete(id); // R#: A possible null value cannot be used for a type marked with [NotNull] or [DisallowNull]
    }

    private void Delete([DisallowNull] TId idToDelete)
    {
        if (idToDelete == null)
        {
            throw new InvalidOperationException("id must not be null.");
        }
    }

    private object GetId()
    {
        return _id!;
    }
}
```
I've added a breakpoint on the `Delete` method invocation. Stepping through the code in the debugger, the value of `id` is actually `null` at the breakpoint.

Because `TId` is unconstrained (so potentially null), the cast at line:
```c#
var id = (TId)GetId();
```
should infer a nullability state of possibly-null for `id`.

Changing the line `Delete(id);` to `Delete(id!);` satisfies ReSharper, but breaks the build with:
```
IDE0370: Suppression is unnecessary
```

**Comment by CyrusNajmabadi** [comment]
Tid is not marked as Nullable in your cast.  So resharper is wrong here and ide370 is correct.

**Comment by bkoelman** [comment]
Thanks. Tracked at https://youtrack.jetbrains.com/issue/RSRP-503026/Incorrect-nullable-analysis-in-casts-to-unconstrained-generics.
